pyproject.toml

```diff
@@ -20,7 +20,8 @@ classifiers = [
   "Programming Language :: Python :: Implementation :: PyPy"
 ]
 dependencies = [
-  "hatch>=1.7.0"
+  "hatch>=1.7.0,<2",
+  "pip-tools>=6,<8"
```

Comment:
I found our usage of `pip-tools` API to be compatible with 6.x and 7.x - 8.x isn't released so I fixed it there. This doesn't affect which version of `pip-tools` gets installed into the virtualenvs and invoked via the API though.

Reply:
Please don't add upper version constraints.
https://iscinumpy.dev/post/bound-version-constraints/

Reply:
Yeah, that was my initial inclination as well - thanks for the gut check. I'll remove the upper bound and keep an eye out for any breaking API changes

Reply:
I see that a private import is used.. that's definitely more scary then 😕
Is there any way to instead copy their code or something

Reply:
The private import is the reason I was so strict on the version pinning. Unfortunately it's private imports all the way down as `pip-tools` uses a bunch of private methods from `pip` https://github.com/jazzband/pip-tools/blob/main/piptools/_compat/pip_compat.py

Ultimately it's `pip._internal.req.parse_requirements` that this PR is leveraging plus some `pip-tools` extras

Reply:
I could trade `pip-tools` private imports for `pip` private imports. Not sure if this makes anything better though.

```python
from packaging.requirements import Requirement
from pip._internal.network.session import PipSession
from pip._internal.req import parse_requirements
from pip._internal.req.constructors import install_req_from_parsed_requirement


def lockfile_to_requirements(lockfile: str) -> list[Requirement]:
    requirements = []
    session = PipSession()
    for parsed_req in parse_requirements(
        lockfile, session, finder=None, options=None, constraint=False
    ):
        install_req = install_req_from_parsed_requirement(parsed_req, isolated=False)
        requirements.append(install_req.req)
    return requirements
```

Reply:
Yeah nothing to do about imports then.

I was just checking that pip-tools doesn't pin pip's version. But of course that makes sense, pinning it could cause unimaginable issues

Reply:
They enforce `>= 22.2` https://github.com/jazzband/pip-tools/blob/e02d186efc74bd2509b4a0f038b0184bdc106e42/pyproject.toml#L41

So just to confirm here, you're recommending to go forward with the private piptools import and to remove the upper bound dependency?

Reply:
I think so..